### PR TITLE
Custom Training extensions for ASR chainer

### PR DESCRIPTION
**Pull request by Fhrozen**
Given that some lines does not fit for transformer (subsampling, truncated_backprop, gpu variables from the converter, etc), I separated the CustomUpdater, ..., Converter.

These has been moved to .training file.
There is no much effect in rnn model. For transformer the forwarding time is reduced by 25%, specially since most of the initial variables are preprocessed in the cpu before moving to the gpu.

### CODE REVIEW

**Comment by Fhrozen** [comment]
@sw005320 The fails on CircleCI are because of that conda could not be downloaded.
Do you mind restarting the tests?

File: espnet/nets/chainer_backend/transformer/training.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+    def __call__(self, batch, device):
+        """Perform sabsampling.
+
+        Args:
+            batch (list): Batch that will be sabsampled.
+            device (device): GPU device.
+
+        Returns:
+            chainer.Variable: xp.array that sabsampled from batch.
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/training.py]:
```suggestion
            chainer.Variable: xp.array that are padded and subsampled from batch.
```

File: espnet/nets/chainer_backend/transformer/training.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            xp.array: xp.array of the length of the mini-batches.
+            chainer.Variable: xp.array that sabsampled from batch.
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/training.py]:
```suggestion
            chainer.Variable: xp.array that are padded and subsampled from batch.
```

File: espnet/nets/chainer_backend/transformer/training.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    def __init__(self, subsampling_factor=1):
+        pass
+
+    def __call__(self, batch, device):
+        """Perform sabsampling.
+
+        Args:
+            batch (list): Batch that will be sabsampled.
+            device (device): GPU device.
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/training.py]:
```suggestion
            device (chainer.backend.Device): CPU or GPU device.
```

File: espnet/nets/chainer_backend/transformer/training.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    Args:
+        subsampling_factor (int): The subsampling factor.
+
+    """
+
+    def __init__(self, subsampling_factor=1):
+        pass
+
+    def __call__(self, batch, device):
+        """Perform sabsampling.
```
Comment (ShigekiKarita) [inline espnet/nets/chainer_backend/transformer/training.py]:
```suggestion
        """Perform subsampling.
```

**Comment by ShigekiKarita** [comment]
I expect that there might be a few conflicts with #943. In my opinion, I want to merge this first because this PR is smaller.

**Pull request merged**